### DISCUSSION TITLE
How to optimize of percentage inputs

### DISCUSSION THREAD

**Discussion by taikichi-pc** [meta-pytorch/botorch#1301] [Q&A] [answered]
When I optimize three inputs (X_bounds = [0-100,  0-100,  0-100]),
Where should I constrain above condition?
Because sum of the three inputs must be just 100.

Please let me know, **one or two example codes** and **reference links**.

I have to write code applying both of single- and Multi-objects.

And this case was used before, I'd like to know the possibility.
When I experimented with them once, the calculations didn't finish.

Thank you for your kindness.

**Comment by saitcakmak** [comment, marked as answer]
Hi @taikichi-pc. [`optimize_acqf`](https://github.com/pytorch/botorch/blob/main/botorch/optim/optimize.py#L81-L83) takes in an `equality_constraints` argument, though there seems to be a bug with this: #1227. 

In your case, the constraint is quite simple, so you could do define a simple wrapper (a sub-class in this case) around the acquisition function, which takes in 2-dimensional inputs and appends the last input before evaluating the acquisition function. You'd then optimize this via `optimize_acqf`, using the bounds for the first two dimensions and using an `inequality_constraint` that makes sure that the sum of the two is `<= 100`. Something like:
```
class WrappedAcqf(qNoisyExpectedImprovement):  # replace with the acquisition function you want to use
    def forward(self, X: Tensor) -> Tensor:
        # X here is a batch x q x 2 dim tensor, denoting only the first two inputs
        third_input = 100 - X.sum(dim=-1, keepdim=True)  # batch x q x 1 dim tensor denoting the third input
        X_full = torch.cat([X, third_input], dim=-1)
        return super().forward(X_full)

model = ...  # create & fit your model
# It is recommended to use input_transform=Normalize(d=3 # model input dimension) & 
# outcome_transform=Standardize(m=# of outcomes / objectives) for better model performance.
acqf = WrappedAcqf(...)  # crate the acqf

limited_bounds = torch.tensor([[0.0, 0.0], [100.0, 100.0]])
candidates, values = optimize_acqf(
    acq_function=acqf,
    bounds=limited_bounds,
    ....  # other arguments,
    inequality_constraints=[(torch.range(2), -torch.ones(2), -100.0)]  # denoting -X[..., 0] - X[..., 1] >= -100
)

```

**Reply by taikichi-pc** [reply]
Hi, saitcakmak -san.
Thank you very much for your quick reply and easy to understand explain.

It does make sense to me.

I love Botorch!
